java-frontend/src/main/java/org/sonar/java/se/ProgramState.java

```diff
@@ -259,6 +260,22 @@ public String toString() {
     return "{" + values.toString() + "}  {" + constraints.toString() + "}" + " { " + stack.toString() + " }";
   }
 
+  public ProgramState addConstraintTransitively(SymbolicValue symbolicValue, Constraint constraint) {
+    final ProgramState[] ps = {addConstraint(symbolicValue, constraint)};
+    knownRelations()
+      .filter(rsv -> rsv.isEquality() && (rsv.getLeftOp() == symbolicValue || rsv.getRightOp() == symbolicValue))
+      .map(rsv -> rsv.getLeftOp() == symbolicValue ? rsv.getRightOp() : rsv.getLeftOp())
+      .forEach(sv -> ps[0] = ps[0].addConstraint(sv, constraint));
```

Comment:
I am pretty sure what we are looking for here is a collector acting as reduction : https://docs.oracle.com/javase/tutorial/collections/streams/reduction.html

While this final array works it is not really nice.